### PR TITLE
WIP: Record import

### PR DESCRIPTION
**Pull request by tjerman**
Allow record import from structured files.

### CODE REVIEW

File: api/compose/spec.json [hunk 1]
```diff
@@ -661,6 +661,66 @@
...
+            },
+            {
+                "name": "importRun",
+                "path": "/import",
+                "method": "PATCH",
+                "title": "Run record import",
+                "parameters": {
+                    "post": [
+                        {
+                            "name": "sessionID",
```
Comment (darh) [inline api/compose/spec.json]: move this to path, so `/import/{sessionID}`

File: api/compose/spec.json [hunk 2]
```diff
@@ -661,6 +661,66 @@
...
+            },
+            {
+                "name": "importProgress",
+                "path": "/import",
+                "method": "GET",
+                "title": "Get import progress",
+                "parameters": {
+                    "get": [
+                        {
+                            "name": "sessionID",
```
Comment (darh) [inline api/compose/spec.json]: move this to path, so `/import/{sessionID}`

File: compose/internal/service/access_control.go [hunk 3]
```diff
@@ -170,6 +170,17 @@ func (svc accessControl) CanRunAutomationTrigger(ctx context.Context, r *automat
 	return svc.can(ctx, types.AutomationScriptPermissionResource.AppendID(r.ID), "run", permissions.Allowed)
 }
 
+// @todo actual controll
+func (svc accessControl) CanReadImportSession(ctx context.Context) bool {
```
Comment (darh) [inline compose/internal/service/access_control.go]: No need for AC for the record importer.
If the user can create records (on module) she can import them.
Session access control is done solely on ownership.

File: compose/rest/record.go [hunk 4]
```diff
@@ -167,6 +180,108 @@ func (ctrl *Record) Upload(ctx context.Context, r *request.RecordUpload) (interf
...
+	// determine decoder
+	switch strings.ToLower(ext) {
+	case "json", "jsonl", "ldjson", "ndjson":
+		recordDecoder = decoder.NewStructuredDecoder(json.NewDecoder(f))
+
+	case "csv":
+		recordDecoder = decoder.NewFlatReader(csv.NewReader(f))
+
+	// @todo...
+	case "xlss":
```
Comment (darh) [inline compose/rest/record.go]: `xlsx`?

File: compose/internal/service/session.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	}
+
+	sessionAccessController interface {
+		CanReadImportSession(context.Context) bool
+		CanCreateImportSession(context.Context) bool
+		CanDeleteImportSession(context.Context) bool
+	}
+
+	SessionService interface {
+		With(ctx context.Context) SessionService
```
Comment (darh) [inline compose/internal/service/session.go]: Please do not follow this pattern (`With(context.Context)`) anymore. This was a design mistake. Will be refactored soon.

> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx

More: https://golang.org/pkg/context/

See how AccessControl, AutomationScript, AutomationTrigger are done.

File: compose/decoder/decoder.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+	flatReader struct {
+		r      FlatReader
+		header []string
+		more   bool
+	}
+
+	structuredDecoder struct {
+		header []string
+		d      StructuredDecoder
+		buf    []map[string]interface{}
```
Comment (darh) [inline compose/decoder/decoder.go]: We're storing all entries from the file in this buffer? Do I understand this correctly?

If yes -- this is not how this should be implemented, no reason to hold it in a buffer.

File: compose/decoder/decoder.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+	dec.Get(func(rtr []string) {
+		dec.header = rtr
+	})
+
+	return dec.header
+}
+
+// structured decoder
+func NewStructuredDecoder(d StructuredDecoder) *structuredDecoder {
```
Comment (darh) [inline compose/decoder/decoder.go]: You'll probably need decoder (for converting one file line to structure, something like `json.Decoder`) AND reader (io.Reader). Then, structured decoder should pass reader through something that knows how to stop at new-line and feed what was read into decoder.

File: compose/rest/record.go [hunk 8]
```diff
@@ -167,6 +180,108 @@ func (ctrl *Record) Upload(ctx context.Context, r *request.RecordUpload) (interf
...
+
+	// Make session
+	ses := types.RecordImportSession{
+		SessionID:    factory.Sonyflake.NextID(),
+		ImportFields: recordDecoder.Header(),
+		NamespaceID:  r.NamespaceID,
+		ModuleID:     r.ModuleID,
+	}
+
+	ctrl.session.SetByID(ses.SessionID, &recordImport{decoder: recordDecoder, session: ses})
```
Comment (darh) [inline compose/rest/record.go]: 1. Move the responsibility of generating session-IDs out of controller into service
2. passing non-exported structs and accept it as interface{}, not something I would be proud of. Pass params in as directly as possible (along with decoder)

File: compose/internal/service/service.go [hunk 9]
```diff
@@ -48,11 +48,12 @@ var (
 	DefaultAutomationRunner automationRunner
 
 	DefaultNamespace NamespaceService
+	DefaultSession   SessionService
 	DefaultRecord    RecordService
 	DefaultModule    ModuleService
 	DefaultChart     ChartService
 	DefaultPage      PageService
-
+	// DefaultTrigger      TriggerService
```
Comment (darh) [inline compose/internal/service/service.go]: This looks like a relic from before latest automation feature. Remove.

File: compose/internal/service/session.go [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+type (
+	session struct {
```
Comment (darh) [inline compose/internal/service/session.go]: This is too general, rename to `ImportSession`

File: compose/internal/service/record.go [hunk 11]
```diff
@@ -184,6 +218,49 @@ func (svc record) Find(filter types.RecordFilter) (set types.RecordSet, f types.
...
+	if ses.Progress.StartedAt != nil {
+		return errors.New("Unable to start import: Import session already active")
+	}
+
+	sa := time.Now()
+	ses.Progress.StartedAt = &sa
+	ssvc.SetRecordByID(svc.ctx, ses.SessionID, 0, 0, nil, &ses.Progress, nil)
+
+	return svc.db.Transaction(func() (err error) {
+		err = (*ses.Decoder).Records(ses.Fields, func(mod *types.Record) error {
```
Comment (tjerman) [inline compose/internal/service/record.go]: @darh is something like this (`(*ses.Decoder).Records`) ok; it doesn't seem ok to me

File: compose/internal/service/service.go [hunk 12]
```diff
@@ -130,6 +130,7 @@ func Init(ctx context.Context, log *zap.Logger, c Config) (err error) {
 		)
 	}
 
+	DefaultSession = ImportSession()
```
Comment (darh) [inline compose/internal/service/service.go]: => `DefaultImportSession`

File: compose/internal/service/record.go (hunk 11 above)
Comment (darh) [reply compose/internal/service/record.go]: nope.

File: pkg/helpers/files.go [hunk 13]
```diff
@@ -0,0 +1,54 @@
+package helpers
```
Comment (darh) [inline pkg/helpers/files.go]: https://blog.golang.org/package-names

nope.

How about just mime?

File: compose/rest/record.go [hunk 14]
```diff
@@ -167,6 +171,106 @@ func (ctrl *Record) Upload(ctx context.Context, r *request.RecordUpload) (interf
...
+	switch strings.ToLower(ext) {
+	case "json", "jsonl", "ldjson", "ndjson":
+		recordDecoder = decoder.NewStructuredDecoder(json.NewDecoder(f))
+
+	case "csv":
+		recordDecoder = decoder.NewFlatReader(csv.NewReader(f))
+
+	// @todo...
+	case "xlsx":
+		// recordDecoder = encoder.NewExcelizeDecoder()
```
Comment (darh) [inline compose/rest/record.go]: nil pointer error.

return error - not implemented or something.

File: compose/internal/service/import_session.go [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+	var ris *RecordImportSession
+
+	if i >= 0 {
+		ris = svc.records[i]
+	} else {
+		ris = &RecordImportSession{
+			SessionID: factory.Sonyflake.NextID(),
+			CreatedAt: time.Now(),
+		}
+		svc.records = append(svc.records, ris)
```
Comment (darh) [inline compose/internal/service/import_session.go]: This (modifying same variable from multiple goroutines) is not safe.
Add locking (see internal/permissions/service.go for examples)

File: compose/internal/service/service.go (hunk 12 above)
Comment (tjerman) [reply compose/internal/service/service.go]: Ee... slipped by

**Pull request merged**